### PR TITLE
hide issues from org private repos w/o team assignment

### PR DESCRIPTION
**Pull request by daviian**
targets https://github.com/go-gitea/gitea/issues/4029

hides private organization repository issues from users that have no access to them.

### CODE REVIEW

File: models/access_test.go [hunk 1]
```diff
@@ -47,8 +49,10 @@ func TestHasAccess(t *testing.T) {
 
 	user1 := AssertExistsAndLoadBean(t, &User{ID: 2}).(*User)
 	user2 := AssertExistsAndLoadBean(t, &User{ID: 5}).(*User)
-	repo1 := AssertExistsAndLoadBean(t, &Repository{OwnerID: 2, IsPrivate: false}).(*Repository)
-	repo2 := AssertExistsAndLoadBean(t, &Repository{OwnerID: 3, IsPrivate: true}).(*Repository)
+	// A public repository owned by User 2
+	repo1 := AssertExistsAndLoadBean(t, &Repository{ID: 1}).(*Repository)
```
Comment (jonasfranz) [inline models/access_test.go:55]: Why did you remove the `IsPrivate: false` part?

File: models/access_test.go [hunk 2]
```diff
@@ -47,8 +49,10 @@ func TestHasAccess(t *testing.T) {
...
+	// A private repository owned by Org 3
+	repo2 := AssertExistsAndLoadBean(t, &Repository{ID: 3}).(*Repository)
```
Comment (jonasfranz) [inline models/access_test.go]: Why did you remove the `IsPrivate: true` part?

File: models/access_test.go [hunk 3]
```diff
@@ -22,8 +22,10 @@ func TestAccessLevel(t *testing.T) {
...
```
Comment (jonasfranz) [inline models/access_test.go]: Why did you remove the `IsPrivate: true` part?

File: models/access_test.go [hunk 4]
```diff
@@ -22,8 +22,10 @@ func TestAccessLevel(t *testing.T) {
 
 	user1 := AssertExistsAndLoadBean(t, &User{ID: 2}).(*User)
 	user2 := AssertExistsAndLoadBean(t, &User{ID: 5}).(*User)
-	repo1 := AssertExistsAndLoadBean(t, &Repository{OwnerID: 2, IsPrivate: false}).(*Repository)
-	repo2 := AssertExistsAndLoadBean(t, &Repository{OwnerID: 3, IsPrivate: true}).(*Repository)
+	// A public repository owned by User 2
+	repo1 := AssertExistsAndLoadBean(t, &Repository{ID: 1}).(*Repository)
```
Comment (jonasfranz) [inline models/access_test.go:26]: Why did you remove the `IsPrivate: false` part?

File: models/user.go [hunk 5]
```diff
@@ -557,7 +557,9 @@ func (u *User) GetOrgRepositoryIDs() ([]int64, error) {
 	var ids []int64
 	return ids, x.Table("repository").
 		Cols("repository.id").
-		Join("INNER", "team_user", "repository.owner_id = team_user.org_id AND team_user.uid = ?", u.ID).
+		Join("INNER", "team_user", "repository.owner_id = team_user.org_id").
+		Join("INNER", "team_repo", "(team_user.team_id = team_repo.team_id AND repository.id = team_repo.repo_id) OR repository.is_private = 0").
```
Comment (jonasfranz) [inline models/user.go]: Check `is_private = 0` first

File: models/user_test.go [hunk 6]
```diff
@@ -159,3 +159,26 @@ func BenchmarkHashPassword(b *testing.B) {
...
+	accessibleRepos, err = user4.GetOrgRepositoryIDs()
+	assert.NoError(t, err)
+	// User 4's team has access to private repo 3, repo 32 is a public repo of the organization
+	assert.Equal(t, accessibleRepos, []int64{3, 32})
+
+	accessibleRepos, err = user5.GetOrgRepositoryIDs()
+	assert.NoError(t, err)
+	// User 5's team has no access to any repo
+	var emptyRepos []int64
+	assert.Equal(t, accessibleRepos, emptyRepos)
```
Comment (jonasfranz) [inline models/user_test.go]: Please use `assert.Len` instead

**Review by jonasfranz — CHANGES_REQUESTED** [review]

**Comment by jonasfranz** [comment]
[Drone ci](https://drone.gitea.io/go-gitea/gitea/928/10) is failing.

File: models/user_test.go [hunk 7]
```diff
@@ -159,3 +159,26 @@ func BenchmarkHashPassword(b *testing.B) {
...
+func TestGetOrgRepositoryIDs(t *testing.T) {
+	assert.NoError(t, PrepareTestDatabase())
+	user2 := AssertExistsAndLoadBean(t, &User{ID: 2}).(*User)
+	user4 := AssertExistsAndLoadBean(t, &User{ID: 4}).(*User)
+	user5 := AssertExistsAndLoadBean(t, &User{ID: 5}).(*User)
+
+	accessibleRepos, err := user2.GetOrgRepositoryIDs()
+	assert.NoError(t, err)
+	// User 2's team has access to private repos 3, 5, repo 32 is a public repo of the organization
+	assert.Equal(t, accessibleRepos, []int64{3, 5, 32})
```
Comment (jonasfranz) [inline models/user_test.go]: The second parameter should be the expected value and the third the actual value. (See https://godoc.org/github.com/stretchr/testify/assert#Equal)

File: models/user_test.go [hunk 8]
```diff
@@ -159,3 +159,26 @@ func BenchmarkHashPassword(b *testing.B) {
...
+
+	accessibleRepos, err = user4.GetOrgRepositoryIDs()
+	assert.NoError(t, err)
+	// User 4's team has access to private repo 3, repo 32 is a public repo of the organization
+	assert.Equal(t, accessibleRepos, []int64{3, 32})
```
Comment (jonasfranz) [inline models/user_test.go]: The second parameter should be the expected value and the third the actual value. (See https://godoc.org/github.com/stretchr/testify/assert#Equal)

**Review by jonasfranz — CHANGES_REQUESTED** [review]

File: models/access_test.go (hunk 1 above)
Comment (jonasfranz) [reply models/access_test.go:55]: > [13:16] lunny: JonasFranz because bool has default value false, Don’t know it means false or null
> [13:16] lunny: Sess.Get has the same behavior
> [13:16] JonasFranz: Okay, so we need to check it afterwards?

Since it is not supported to use bools in xorm, I propose to add an check if IsPrivate is set right:

```go
assert.False(t, repo1.IsPrivate)
```
or

```go
assert.True(t, repo1.IsPrivate)
```

Comment (daviian) [reply models/access_test.go:55]: I changed it such that it really loads the expected repository. The resulting repository is the same, just an assurance.
That's applicable to the other changes in `access_test.go` too.

Comment (jonasfranz) [reply models/access_test.go:55]: But it would be good to check that the repository is really private/public a comment does not ensure that.

File: models/user.go (hunk 5 above)
Comment (JonasFranzDEV) [reply models/user.go]: This is not fixed. Please change it to: 
`repository.is_private = 0 OR (team_user.team_id = team_repo.team_id AND repository.id = team_repo.repo_id)`

File: models/user.go [hunk 9]
```diff
@@ -557,7 +557,9 @@ func (u *User) GetOrgRepositoryIDs() ([]int64, error) {
 	var ids []int64
 	return ids, x.Table("repository").
 		Cols("repository.id").
-		Join("INNER", "team_user", "repository.owner_id = team_user.org_id AND team_user.uid = ?", u.ID).
+		Join("INNER", "team_user", "repository.owner_id = team_user.org_id").
+		Join("INNER", "team_repo", "(team_user.team_id = team_repo.team_id AND repository.id = team_repo.repo_id) OR repository.is_private != ?", true).
```
Comment (jonasfranz) [inline models/user.go]: Please change it to:
```sql
repository.is_private = 0 OR (team_user.team_id = team_repo.team_id AND repository.id = team_repo.repo_id)
```

**Review by jonasfranz — CHANGES_REQUESTED** [review]

File: models/user.go (hunk 9 above)
Comment (daviian) [reply models/user.go]: Problem with this is that it's not necessarily `0` if it is not private, because it could also be `null`
Also integration tests with pgsql fail on my pc

File: models/access_test.go (hunk 1 above)
Comment (daviian) [reply models/access_test.go:55]: I can add it of course, just think that it's nothing that adds value since its predefined by the fixtures anyway.

File: models/user.go [hunk 10]
```diff
@@ -558,7 +558,7 @@ func (u *User) GetOrgRepositoryIDs() ([]int64, error) {
 	return ids, x.Table("repository").
 		Cols("repository.id").
 		Join("INNER", "team_user", "repository.owner_id = team_user.org_id").
-		Join("INNER", "team_repo", "(team_user.team_id = team_repo.team_id AND repository.id = team_repo.repo_id) OR repository.is_private != ?", true).
```
Comment (jonasfranz) [inline models/user.go]: Please remove `, true` since it is not used anymore

**Review by jonasfranz — CHANGES_REQUESTED** [review]

File: models/user.go (hunk 10 above)
Comment (daviian) [reply models/user.go]: Actually it is used. It's not possible to write `!= 0` because it doesn't work on all databases. You will see that pgsql fails if we use that.

**Review by jonasfranz — APPROVED** [review]

**Review by jonasfranz — APPROVED** [review]

**Review by jonasfranz — CHANGES_REQUESTED** [review]
This may be out of scope for this PR but if I create a team with permissions for Wiki only for example, the issues of the repository are shown in the issue list of the team members. But they shouldn't see them.

**Comment by lunny** [comment]
LGTM

**Comment by daviian** [comment]
@lunny @JonasFranzDEV Do you want me to make a more or less dirty fix by iterating over the selected repo ID's and check the unittype or do you favor a more solid solution by moving out the UnitType json into a new team_unit table?

**Comment by lafriks** [comment]
I would prefer team_unit table but than it can not be backported

**Comment by daviian** [comment]
@lafriks I could backport the quick'n'dirty solution anyway ;-)

**Comment by lafriks** [comment]
@daviian it is not so critical bug to be worth it imho

**Comment by daviian** [comment]
Can you mark this PR as WIP? There is one thing missing I've not implemented yet.

TODO
- [x] Migrate team unit types into new `team_unit` table
- [x] Filter Issues by team access and unit type
- [x] Don't notify users about certain actions that lack the required unit type.
- [ ] ~Remove link to issue in actions on dashboard if user is not allowed to access~

**Comment by daviian** [comment]
@lafriks There's one open issue though. What happens to existing actions on the dashboard that are linked to unallowed issues because of a change in the units afterwards?

**Comment by lafriks** [comment]
I don't think anything can be done there

**Comment by daviian** [comment]
@lafriks I was thinking about removing the link to the issue if he lacks permission.

**Comment by lafriks** [comment]
That can be done of course but we don't know exact time from when to remove actions, ex. if user had rights before. But removing all actions with no appropriate rights would probably be ok

**Comment by daviian** [comment]
@lafriks Do you know a cleaner way to check user units per action in feed?

**Comment by lafriks** [comment]
I don't think showing link is them problem, why would not show it? You will get 404 if you have no rights anyway, so I don't see problem in that. I don't think that user actions even should have rights checked at all. I was thinking you are asking about actions being created wrongly and need to be deleted now. But I don't think it is so necessary also.

**Comment by daviian** [comment]
Oh ok. No I meant to disable the not working links after disabling it.
So should I undo the last commit? Disabling the links is just more convenient for the user instead of a trial and error approach IMO

**Comment by lafriks** [comment]
imho it is not needed and just complicates things

**Comment by daviian** [comment]
@go-gitea/mergers  Can you please restart drone. Tests are fine locally.

**Comment by jonasfranz** [comment]
What's the status of this PR?

**Comment by daviian** [comment]
All of my todos are finished, see https://github.com/go-gitea/gitea/pull/4034#issuecomment-394183230
Last todo was cancelled because after discussion was assumed not necessary.

So it's ready to be reviewed again and hopefully merged.

File: models/migrations/v67.go [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+	// Team see models/team.go
+	type Team struct {
+		ID        int64 `xorm:"pk autoincr"`
+		UnitTypes []int `xorm:"json"`
+	}
+
+	// TeamUnit see models/org_team.go
+	type TeamUnit struct {
+		ID     int64 `xorm:"pk autoincr"`
+		TeamID int64 `xorm:"INDEX(s)"`
```
Comment (lunny) [inline models/migrations/v67.go]: should be `Unique(s)`

File: models/org_team.go [hunk 12]
```diff
@@ -646,3 +659,46 @@ func GetTeamsWithAccessToRepo(orgID, repoID int64, mode AccessMode) ([]*Team, er
...
+	TeamID int64    `xorm:"INDEX(s)"`
+	Type   UnitType `xorm:"INDEX(s)"`
+}
+
+// Unit returns Unit
+func (t *TeamUnit) Unit() Unit {
+	return Units[t.Type]
+}
+
+func getUnitsByTeamID(e Engine, teamID int64) (units []*TeamUnit, err error) {
```
Comment (lunny) [inline models/org_team.go:739]: I would like this is 
```Go
func getUnitsByTeamID(e Engine, teamID int64) (units []UnitType, err error) {
     return units, e.Table("team_unit").Cols("type").Where("team_id = ?", teamID).Find(&units)
}
```
So that we could not change many `[]UnitType` to `[]*TeamUnit`.

**Review by lunny — CHANGES_REQUESTED** [review]

File: models/org_team.go (hunk 12 above)
Comment (daviian) [reply models/org_team.go:739]: I wanted to be consistent with the RepoUnit method, it returns `[]*RepoUnit`

**Comment by daviian** [comment]
@lunny Btw it doesn't make sense to set `TeamID` to unique since there are as many entries as enabled unit types. So I revert my change of the `unique` index.

**Comment by jonasfranz** [comment]
@daviian Do you delete the related `TeamUnit` after a team gets deleted?

**Comment by daviian** [comment]
@JonasFranzDEV Good catch. 😲

**Comment by lafriks** [comment]
@lunny @JonasFranzDEV need your approval

**Review by lunny — APPROVED** [review]

**Review by jonasfranz — APPROVED** [review]
conflicted but otherwise LGTM

**Comment by lafriks** [comment]
@daviian please rebase and resolve conflicts

**Comment by axifive** [comment]
@daviian, Please use v69 migration name for conflict resolving. Because I already resolved https://github.com/go-gitea/gitea/pull/4258 with v68

**Review by sapk — APPROVED** [review]

**Comment by axifive** [comment]
@daviian, just wanted to warn against double rebase
